<a href="https://colab.research.google.com/github/EllieMwangi/cabdata_analysis/blob/master/Uber_Lyft_data_descriptive_analysis_report.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Load libraries
import pandas as pd
import numpy as np
import datetime as dt
import seaborn as sns

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [ ]:
# Load sql extension
%load_ext sql
%sql sqlite://

'Connected: @None'

In [ ]:
# Load data
car_df = pd.read_csv('https://bit.ly/cabsdataset', infer_datetime_format=True)
weather_df = pd.read_csv('https://bit.ly/cabsweatherdata', infer_datetime_format=True)

In [ ]:
# First five row
car_df.head()

Unnamed: 0  distance  ...    product_id          name
0           0      0.44  ...     lyft_line        Shared
1           1      0.44  ...  lyft_premier           Lux
2           2      0.44  ...          lyft          Lyft
3           3      0.44  ...   lyft_luxsuv  Lux Black XL
4           4      0.44  ...     lyft_plus       Lyft XL

[5 rows x 11 columns]

In [ ]:
# Last five rows
car_df.tail()

Unnamed: 0  distance  ...                            product_id       name
693066      693066       1.0  ...  6f72dfc5-27f1-42e8-84db-ccc7a75f6969     UberXL
693067      693067       1.0  ...  55c66225-fbe7-4fd5-9072-eab1ece5e23e      UberX
693068      693068       1.0  ...  8cf7e821-f0d3-49c6-8eba-e679c0ebcf6a       Taxi
693069      693069       1.0  ...  6d318bcc-22a3-4af6-bddd-b409bfce1546  Black SUV
693070      693070       1.0  ...  997acbb5-e102-41e1-b155-9df7de0a73f2   UberPool

[5 rows x 11 columns]

In [ ]:
# Sample ten records
car_df.sample(10)

Unnamed: 0  distance  ...                            product_id       name
74413        74413      3.44  ...  9a0e7b09-b92b-4c41-9779-2ad22b4d779d        WAV
531738      531738      0.54  ...  6f72dfc5-27f1-42e8-84db-ccc7a75f6969     UberXL
133262      133262      1.25  ...  997acbb5-e102-41e1-b155-9df7de0a73f2   UberPool
265713      265713      3.01  ...                          lyft_premier        Lux
218594      218594      1.30  ...  9a0e7b09-b92b-4c41-9779-2ad22b4d779d        WAV
214234      214234      1.25  ...  6f72dfc5-27f1-42e8-84db-ccc7a75f6969     UberXL
543346      543346      2.79  ...  8cf7e821-f0d3-49c6-8eba-e679c0ebcf6a       Taxi
408839      408839      4.61  ...                          lyft_premier        Lux
197139      197139      2.17  ...                             lyft_line     Shared
50413        50413      3.61  ...  6d318bcc-22a3-4af6-bddd-b409bfce1546  Black SUV

[10 rows x 11 columns]

In [ ]:
# Number of rows and columns
car_df.shape

(693071, 11)

In [ ]:
# Column data types
car_df.dtypes

Unnamed: 0            int64
distance            float64
cab_type             object
time_stamp           object
destination          object
source               object
price               float64
surge_multiplier    float64
id                   object
product_id           object
name                 object
dtype: object

In [ ]:
#Drop unrequired unnamed column
car_df = car_df.drop(['Unnamed: 0'], axis=1)

In [ ]:
car_df.head()

distance cab_type  ...    product_id          name
0      0.44     Lyft  ...     lyft_line        Shared
1      0.44     Lyft  ...  lyft_premier           Lux
2      0.44     Lyft  ...          lyft          Lyft
3      0.44     Lyft  ...   lyft_luxsuv  Lux Black XL
4      0.44     Lyft  ...     lyft_plus       Lyft XL

[5 rows x 10 columns]

In [ ]:
# First five columns of weather dataframe
weather_df.head()

Unnamed: 0   temp            location  ...           time_stamp  humidity   wind
0           0  42.42            Back Bay  ...  2018-12-16 23:45:01      0.77  11.25
1           1  42.43         Beacon Hill  ...  2018-12-16 23:45:01      0.76  11.32
2           2  42.50   Boston University  ...  2018-12-16 23:45:01      0.76  11.07
3           3  42.11              Fenway  ...  2018-12-16 23:45:01      0.77  11.09
4           4  43.13  Financial District  ...  2018-12-16 23:45:01      0.75  11.49

[5 rows x 9 columns]

In [ ]:
# Last five columns
weather_df.tail()

Unnamed: 0   temp  ... humidity  wind
6271        6271  44.72  ...     0.96  1.52
6272        6272  44.85  ...     0.96  1.54
6273        6273  44.82  ...     0.96  1.54
6274        6274  44.78  ...     0.96  1.54
6275        6275  44.69  ...     0.96  1.52

[5 rows x 9 columns]

In [ ]:
#Drop unnamed column
weather_df = weather_df.drop(['Unnamed: 0'], axis=1)

In [ ]:
#Sample 10 rows
weather_df.sample(10)

temp            location  clouds  ...           time_stamp  humidity   wind
1497  42.45       North Station    0.45  ...  2018-12-01 18:52:54      0.57   0.93
589   37.93  Financial District    0.45  ...  2018-11-29 04:46:23      0.69   9.44
4125  41.46              Fenway    0.95  ...  2018-11-27 16:15:20      0.83  10.48
3862  40.62              Fenway    0.83  ...  2018-11-28 22:02:08      0.63   9.21
2638  23.76           North End    0.64  ...  2018-12-18 11:45:01      0.54  16.23
2261  39.36    Haymarket Square    0.39  ...  2018-12-16 08:45:01      0.74   8.14
347   40.46    Theatre District    1.00  ...  2018-11-26 06:24:05      0.92   1.35
4804  40.47            Back Bay    1.00  ...  2018-11-26 06:16:16      0.91   1.19
660   29.94         Beacon Hill    0.15  ...  2018-12-18 17:45:01      0.46  13.52
3583  35.28       South Station    0.81  ...  2018-12-18 05:45:02      0.64  11.96

[10 rows x 8 columns]

In [ ]:
weather_df.shape

(6276, 8)

In [ ]:
weather_df.dtypes

temp          float64
location       object
clouds        float64
pressure      float64
rain          float64
time_stamp     object
humidity      float64
wind          float64
dtype: object

In [ ]:
#Check for missing values
car_df.isnull().sum().sort_values()

distance                0
cab_type                0
time_stamp              0
destination             0
source                  0
surge_multiplier        0
id                      0
product_id              0
name                    0
price               55095
dtype: int64

In [ ]:
weather_df.isnull().sum().sort_values(ascending=False)

rain          5382
wind             0
humidity         0
time_stamp       0
pressure         0
clouds           0
location         0
temp             0
dtype: int64

In [ ]:
#Check for duplicates
car_df.duplicated().any()

False

In [ ]:
weather_df.duplicated().any()

False

In [ ]:
#Fill missing values in rain column with 0
weather_df.rain.fillna(0, inplace=True)
#Drop rows with missing price value
car_df.dropna(inplace=True)

In [ ]:
weather_df.isnull().sum()

temp          0
location      0
clouds        0
pressure      0
rain          0
time_stamp    0
humidity      0
wind          0
dtype: int64

In [ ]:
car_df.isnull().sum()

distance            0
cab_type            0
time_stamp          0
destination         0
source              0
price               0
surge_multiplier    0
id                  0
product_id          0
name                0
dtype: int64

In [ ]:
# Check for unique values in cab type column
car_df.cab_type.unique()

array(['Lyft', 'Uber'], dtype=object)

In [ ]:
car_df.product_id.unique()

array(['lyft_line', 'lyft_premier', 'lyft', 'lyft_luxsuv', 'lyft_plus',
       'lyft_lux', '6f72dfc5-27f1-42e8-84db-ccc7a75f6969',
       '6c84fd89-3f11-4782-9b50-97c468b19529',
       '55c66225-fbe7-4fd5-9072-eab1ece5e23e',
       '9a0e7b09-b92b-4c41-9779-2ad22b4d779d',
       '6d318bcc-22a3-4af6-bddd-b409bfce1546',
       '997acbb5-e102-41e1-b155-9df7de0a73f2'], dtype=object)

In [ ]:
car_df.name.unique()

array(['Shared', 'Lux', 'Lyft', 'Lux Black XL', 'Lyft XL', 'Lux Black',
       'UberXL', 'Black', 'UberX', 'WAV', 'Black SUV', 'UberPool'],
      dtype=object)

In [ ]:
# Convert timestamp column to datetime
car_df.time_stamp = pd.to_datetime(car_df.time_stamp)

In [ ]:
car_df.dtypes

distance                   float64
cab_type                    object
time_stamp          datetime64[ns]
destination                 object
source                      object
price                      float64
surge_multiplier           float64
id                          object
product_id                  object
name                        object
dtype: object

In [ ]:
# Extract other columns from timestamp column
car_df['date'] = car_df.time_stamp.dt.date

In [ ]:
car_df['hour'] = car_df.time_stamp.dt.hour

In [ ]:
car_df['day'] = car_df.time_stamp.dt.day_name()

In [ ]:
car_df['time'] = car_df.time_stamp.dt.time

In [ ]:
car_df.head()

distance cab_type              time_stamp  ... hour        day             time
0      0.44     Lyft 2018-12-16 09:30:07.890  ...    9     Sunday  09:30:07.890000
1      0.44     Lyft 2018-11-27 02:00:23.677  ...    2    Tuesday  02:00:23.677000
2      0.44     Lyft 2018-11-28 01:00:22.198  ...    1  Wednesday  01:00:22.198000
3      0.44     Lyft 2018-11-30 04:53:02.749  ...    4     Friday  04:53:02.749000
4      0.44     Lyft 2018-11-29 03:49:20.223  ...    3   Thursday  03:49:20.223000

[5 rows x 14 columns]

In [ ]:
# Extract other columns from timestamp column in weather data
weather_df.time_stamp = pd.to_datetime(weather_df.time_stamp)

In [ ]:
weather_df['date'] = weather_df.time_stamp.dt.date

In [ ]:
weather_df['hour'] = weather_df.time_stamp.dt.hour

In [ ]:
weather_df['day'] = weather_df.time_stamp.dt.day_name()

In [ ]:
weather_df.head()

temp            location  clouds  pressure  ...   wind        date  hour     day
0  42.42            Back Bay     1.0   1012.14  ...  11.25  2018-12-16    23  Sunday
1  42.43         Beacon Hill     1.0   1012.15  ...  11.32  2018-12-16    23  Sunday
2  42.50   Boston University     1.0   1012.15  ...  11.07  2018-12-16    23  Sunday
3  42.11              Fenway     1.0   1012.13  ...  11.09  2018-12-16    23  Sunday
4  43.13  Financial District     1.0   1012.14  ...  11.49  2018-12-16    23  Sunday

[5 rows x 11 columns]

In [ ]:
car_df.sort_values(by=['date','hour'], ascending=False)

distance cab_type  ...      day             time
305         2.16     Lyft  ...  Tuesday  19:10:03.694000
1592        0.53     Lyft  ...  Tuesday  19:15:06.703000
1753        0.46     Lyft  ...  Tuesday  19:05:06.293000
3749        4.44     Uber  ...  Tuesday  19:05:06.489000
6007        0.90     Lyft  ...  Tuesday  19:00:05.360000
...          ...      ...  ...      ...              ...
594575      1.71     Uber  ...   Monday  03:40:47.227000
613926      2.71     Uber  ...   Monday  03:40:46.320000
613927      2.71     Uber  ...   Monday  03:40:46.320000
658078      3.44     Uber  ...   Monday  03:40:47.023000
658079      3.44     Uber  ...   Monday  03:40:47.023000

[637976 rows x 14 columns]

Ride dataset ranges from 26/11/2018 3:40 am to 18/12/2018 19:10 pm

In [ ]:
weather_df.sort_values(by=['date','hour'], ascending=False)

temp            location  clouds  ...        date  hour      day
483   30.80         Beacon Hill    0.00  ...  2018-12-18    18  Tuesday
484   30.96   Boston University    0.00  ...  2018-12-18    18  Tuesday
485   30.93              Fenway    0.00  ...  2018-12-18    18  Tuesday
486   31.19  Financial District    0.00  ...  2018-12-18    18  Tuesday
487   30.83       North Station    0.00  ...  2018-12-18    18  Tuesday
...     ...                 ...     ...  ...         ...   ...      ...
3999  41.02         Beacon Hill    0.87  ...  2018-11-26     3   Monday
4000  41.07   Boston University    0.86  ...  2018-11-26     3   Monday
4001  40.95  Financial District    0.89  ...  2018-11-26     3   Monday
4002  40.94       North Station    0.89  ...  2018-11-26     3   Monday
4003  40.80    Theatre District    0.87  ...  2018-11-26     3   Monday

[6276 rows x 11 columns]

Weather dataset ranges from 26/11/2018 3:40 am to 18/12/2018 18:45 pm

In [ ]:
# Group time periods to parts of the day
car_df['early_morning'] = (car_df.hour >= 5) & (car_df.hour <= 8) 

In [ ]:
car_df.early_morning = car_df.early_morning.apply(np.int)

In [ ]:
car_df['late_morning'] = (car_df.hour >= 9) & (car_df.hour <=11 ) 

In [ ]:
car_df.late_morning = car_df.late_morning.apply(np.int)

In [ ]:
car_df['lunch_hours'] = (car_df.hour >= 12) & (car_df.hour <= 13) 

In [ ]:
car_df.lunch_hours = car_df.lunch_hours.apply(np.int)

In [ ]:
car_df['afternoon'] = (car_df.hour >= 14) & (car_df.hour <= 15) 

In [ ]:
car_df.afternoon = car_df.afternoon.apply(np.int)

In [ ]:
car_df['rush_hours'] = (car_df.hour >= 16) & (car_df.hour <= 19)

In [ ]:
car_df['night'] = (car_df.hour >= 20) | ((car_df.hour >= 0 ) & (car_df.hour <= 4))

In [ ]:
car_df.rush_hours = car_df.rush_hours.apply(np.int)
car_df.night = car_df.night.apply(np.int)


In [ ]:
car_df.sort_values(by=['date','hour'], ascending=False)

distance cab_type              time_stamp  ... afternoon rush_hours  night
305         2.16     Lyft 2018-12-18 19:10:03.694  ...         0          1      0
1592        0.53     Lyft 2018-12-18 19:15:06.703  ...         0          1      0
1753        0.46     Lyft 2018-12-18 19:05:06.293  ...         0          1      0
3749        4.44     Uber 2018-12-18 19:05:06.489  ...         0          1      0
6007        0.90     Lyft 2018-12-18 19:00:05.360  ...         0          1      0
...          ...      ...                     ...  ...       ...        ...    ...
594575      1.71     Uber 2018-11-26 03:40:47.227  ...         0          0      1
613926      2.71     Uber 2018-11-26 03:40:46.320  ...         0          0      1
613927      2.71     Uber 2018-11-26 03:40:46.320  ...         0          0      1
658078      3.44     Uber 2018-11-26 03:40:47.023  ...         0          0      1
658079      3.44     Uber 2018-11-26 03:40:47.023  ...         0          0      1

[637976 rows x 20 columns]

In [ ]:
# Derive other variable
car_df['day_of_week'] = car_df.time_stamp.dt.dayofweek

In [ ]:
#Identify if day is a weekend or not: 0 - 4 weekdays, 5 - 6 weekends
car_df['weekday'] = (car_df.day_of_week >= 0) & (car_df.day_of_week <= 4)
car_df['weekend'] = (car_df.day_of_week >= 5) & (car_df.day_of_week <= 6)

In [ ]:
car_df.weekday = car_df.weekday.apply(np.int)
car_df.weekend = car_df.weekend.apply(np.int)

In [ ]:
car_df.head()

distance cab_type              time_stamp  ... day_of_week weekday  weekend
0      0.44     Lyft 2018-12-16 09:30:07.890  ...           6       0        1
1      0.44     Lyft 2018-11-27 02:00:23.677  ...           1       1        0
2      0.44     Lyft 2018-11-28 01:00:22.198  ...           2       1        0
3      0.44     Lyft 2018-11-30 04:53:02.749  ...           4       1        0
4      0.44     Lyft 2018-11-29 03:49:20.223  ...           3       1        0

[5 rows x 23 columns]

In [ ]:
car_df.surge_multiplier.unique()

array([1.  , 1.25, 2.5 , 2.  , 1.75, 1.5 , 3.  ])

In [ ]:
# Create dataframes to be loaded into sql tables
CarData = car_df.drop(['id','product_id','name'], axis=1)
WeatherData = weather_df

In [ ]:
# Merge Data with Pandas as it may be too computationally expensive with SQL
TripData = pd.merge(CarData, WeatherData, how='inner',left_on=['source','date','hour'],right_on=['location','date','hour'])

In [ ]:
TripData.drop(['location','time_stamp_y','day_y'], axis=1, inplace=True)

In [ ]:
# Load data into sql
%sql drop table if exists TripData
%sql persist TripData
%sql select * from TripData limit 5;

 * sqlite://
Done.
 * sqlite://
 * sqlite://
Done.


[(0, 0.44, 'Lyft', '2018-12-16 09:30:07.890000', 'North Station', 'Haymarket Square', 5.0, 1.0, '2018-12-16', 9, 'Sunday', '09:30:07.890000', 0, 1, 0, 0, 0, 0, 6, 0, 1, 38.46, 0.29, 1022.25, 0.0, 0.76, 7.68),
 (1, 1.33, 'Lyft', '2018-12-16 09:25:09.308000', 'Beacon Hill', 'Haymarket Square', 7.0, 1.0, '2018-12-16', 9, 'Sunday', '09:25:09.308000', 0, 1, 0, 0, 0, 0, 6, 0, 1, 38.46, 0.29, 1022.25, 0.0, 0.76, 7.68),
 (2, 0.53, 'Lyft', '2018-12-16 09:05:09.578000', 'North Station', 'Haymarket Square', 27.5, 1.0, '2018-12-16', 9, 'Sunday', '09:05:09.578000', 0, 1, 0, 0, 0, 0, 6, 0, 1, 38.46, 0.29, 1022.25, 0.0, 0.76, 7.68),
 (3, 2.36, 'Lyft', '2018-12-16 09:55:08.860000', 'Back Bay', 'Haymarket Square', 7.0, 1.0, '2018-12-16', 9, 'Sunday', '09:55:08.860000', 0, 1, 0, 0, 0, 0, 6, 0, 1, 38.46, 0.29, 1022.25, 0.0, 0.76, 7.68),
 (4, 0.62, 'Uber', '2018-12-16 09:55:08.161000', 'West End', 'Haymarket Square', 10.5, 1.0, '2018-12-16', 9, 'Sunday', '09:55:08.161000', 0, 1, 0, 0, 0, 0, 6, 0, 1, 38.46, 0.29, 1022.25, 0.0, 0.76, 7.68)]

## **Data Analysis**

In [ ]:
# Average Price Per Day for both Uber and Lyft
%%sql
select day_x as Day, round(avg(price), 2) as 'Average Price'
from TripData
group by day_x
order by avg(price) desc;

 * sqlite://
Done.


[('Thursday', 16.6),
 ('Tuesday', 16.59),
 ('Sunday', 16.58),
 ('Saturday', 16.55),
 ('Wednesday', 16.5),
 ('Monday', 16.49),
 ('Friday', 16.48)]

In [ ]:
# Average Price Per Day for Uber
%%sql
select day_x as Day, round(avg(price), 2) as 'Average Price'
from TripData
where cab_type == 'Uber'
group by day_x
order by avg(price) desc;

 * sqlite://
Done.


[('Tuesday', 15.88),
 ('Thursday', 15.85),
 ('Sunday', 15.83),
 ('Monday', 15.78),
 ('Saturday', 15.78),
 ('Wednesday', 15.77),
 ('Friday', 15.69)]

In [ ]:
# Average Price Per Day for Lyft
%%sql
select day_x as Day, round(avg(price), 2) as 'Average Price'
from TripData
where cab_type == 'Lyft'
group by day_x
order by avg(price) desc;

 * sqlite://
Done.


[('Thursday', 17.4),
 ('Sunday', 17.38),
 ('Saturday', 17.37),
 ('Tuesday', 17.35),
 ('Friday', 17.32),
 ('Wednesday', 17.28),
 ('Monday', 17.26)]

- For both ride apps, average rates are highest on weekdays, Tuesday for Uber and Thursday for Lyft

In [ ]:
%%sql
create view [Early morning Trips] as
select count(surge_multiplier) as Total
from TripData
where early_morning == 1

 * sqlite://
Done.


[]

In [ ]:
%sql select * from [Early morning Trips];

 * sqlite://
Done.


[(153887,)]

In [ ]:
%%sql
create view [Surged early morning trips] as
select count(surge_multiplier) as Total_Surges
from TripData
where early_morning == 1 and surge_multiplier > 1.0

 * sqlite://
Done.


[]

In [ ]:
%sql select * from [Surged early morning trips]

 * sqlite://
Done.


[(5283,)]

In [ ]:
# Check for frequency of price surges in the early morning hours
%%sql
select ROUND((CAST(Total_Surges AS FLOAT) / Total) * 100, 2)  as 'Trip %'
from [Early morning Trips], [Surged early morning trips]

 * sqlite://
Done.


[(3.43,)]

In [ ]:
# Check for frequency of price surges in the late morning hours
%%sql
create view [Late morning Trips] as
select count(surge_multiplier) as Total
from TripData
where late_morning == 1

 * sqlite://
Done.


[]

In [ ]:
%%sql
create view [Surged late morning trips] as
select count(surge_multiplier) as Total_Surges
from TripData
where late_morning == 1 and surge_multiplier > 1.0

 * sqlite://
Done.


[]

In [ ]:
%%sql
select ROUND((CAST(Total_Surges AS FLOAT) / Total) * 100, 2)  as 'Trip %'
from [Late morning Trips], [Surged late morning trips]

 * sqlite://
Done.


[(3.12,)]

In [ ]:
# Check for frequency of price surges in the lunch hours
%%sql
create view [Lunch Hours] as
select count(surge_multiplier) as Total
from TripData
where lunch_hours == 1

 * sqlite://
Done.


[]

In [ ]:
%%sql
create view [Surged lunch hour trips] as
select count(surge_multiplier) as Total_Surges
from TripData
where lunch_hours == 1 and surge_multiplier > 1.0

 * sqlite://
Done.


[]

In [ ]:
%%sql
select ROUND((CAST(Total_Surges AS FLOAT) / Total) * 100, 2)  as 'Trip %'
from [Lunch Hours], [Surged lunch hour trips]

 * sqlite://
Done.


[(3.59,)]

In [ ]:
# Check for frequency of price surges in the afternoon hours
%%sql
create view [Afternoon Trips] as
select count(surge_multiplier) as Total
from TripData
where afternoon == 1

 * sqlite://
Done.


[]

In [ ]:

%%sql
create view [Surged afternoon trips] as
select count(surge_multiplier) as Total_Surges
from TripData
where afternoon == 1 and surge_multiplier > 1.0

 * sqlite://
Done.


[]

In [ ]:
%%sql
select ROUND((CAST(Total_Surges AS FLOAT) / Total) * 100, 2)  as 'Trip %'
from [Afternoon Trips], [Surged afternoon trips]

 * sqlite://
Done.


[(3.12,)]

In [ ]:
# Check for frequency of price surges in the rush hours
%%sql
create view [Rush hour trips] as
select count() as Total
from TripData 
where rush_hours == 1

 * sqlite://
Done.


[]

In [ ]:
%%sql
create view [Surged rush hour trips] as
select count() as Total_Surges
from TripData 
where rush_hours == 1 and surge_multiplier > 1.0

 * sqlite://
Done.


[]

In [ ]:
%%sql
select ROUND((CAST(Total_Surges AS FLOAT) / Total) * 100, 2)  as 'Trip %'
from [Rush hour trips], [Surged rush hour trips]

 * sqlite://
Done.


[(3.16,)]

In [ ]:
# Check for frequency of price surges in the night hours
%%sql
create view [Night trips] as
select count() as Total
from TripData
where night == 1

 * sqlite://
Done.


[]

In [ ]:
%%sql
create view [Surged night trips] as
select count() as Total_Surges 
from TripData
where night == 1 and surge_multiplier > 1.0

 * sqlite://
Done.


[]

In [ ]:
%%sql
select ROUND((CAST(Total_Surges AS FLOAT) / Total) * 100, 2)  as 'Trip %'
from [Night Trips], [Surged night trips]

 * sqlite://
Done.


[(3.33,)]

In [ ]:
# Check for frequency of price surges in entire dataset
%%sql
select surge_multiplier, count(surge_multiplier) as 'Frequency'
from TripData 
group by surge_multiplier;

 * sqlite://
Done.


[(1.0, 1126594),
 (1.25, 20203),
 (1.5, 9384),
 (1.75, 4538),
 (2.0, 4013),
 (2.5, 243),
 (3.0, 21)]

Night, early morning and lunch hour trips have a detected higher percentage of trips with a price surge:
- Lunch hours [ 12:00 pm - 2:00 pm ] : 3.59% of trips have a price surge
- Early morning hours [ 5:00 am - 10:00 am] : 3.43% of trips have a price surge
- Night hours [ 8:00 pm - 4:00 am] : 3.33% of trips have a price surge

In [ ]:
# What are the average weather conditions for each surge multiplier?
%%sql
select round(avg(temp), 2) as 'Average Temperature', round(avg(rain),2) as 'Average Rainfall', round(avg(clouds),2) as 'Cloud Coverage', surge_multiplier
from TripData 
group by surge_multiplier;


 * sqlite://
Done.


[(39.26, 0.01, 0.68, 1.0),
 (39.2, 0.01, 0.68, 1.25),
 (39.06, 0.01, 0.68, 1.5),
 (39.37, 0.01, 0.69, 1.75),
 (39.22, 0.01, 0.67, 2.0),
 (38.98, 0.0, 0.65, 2.5),
 (38.81, 0.01, 0.8, 3.0)]

Trips that experienced price surges, i.e 3.0, have a notable spike in cloud coverage: 0.8

In [ ]:
# what are the average prices on weekdays and weekends?
%%sql
select round(avg(price), 2) as 'Average Price on Weekdays'
from TripData
where weekday == 1

 * sqlite://
Done.


[(16.54,)]

In [ ]:
%%sql
select round(avg(price), 2) as 'Average Price on Weekends'
from TripData
where weekend == 1

 * sqlite://
Done.


[(16.56,)]

Cab prices are relatively equal for weekends and weekdays

In [ ]:
%%sql
select round(avg(price), 2) as 'Average Price on Weekdays'
from TripData
where weekday == 1 and cab_type == 'Uber'

 * sqlite://
Done.


[(15.81,)]

In [ ]:
%%sql
select round(avg(price), 2) as 'Average Price on Weekends'
from TripData
where weekend == 1 and cab_type == 'Uber';

 * sqlite://
Done.


[(15.8,)]

In [ ]:
%%sql
select round(avg(price), 2) as 'Average Price on Weekdays'
from TripData
where weekday == 1 and cab_type == 'Lyft'

 * sqlite://
Done.


[(17.32,)]

In [ ]:
%%sql
select round(avg(price), 2) as 'Average Price on Weekends'
from TripData
where weekend == 1 and cab_type == 'Lyft'

 * sqlite://
Done.


[(17.38,)]

In [ ]:
# What are the average prices by time of day
%%sql
select round(avg(price),2) as 'Average Price'
from TripData
where early_morning == 1

 * sqlite://
Done.


[(16.55,)]

In [ ]:
%%sql
select round(avg(price),2) as 'Average Price'
from TripData
where late_morning == 1

 * sqlite://
Done.


[(16.48,)]

In [ ]:
%%sql
select round(avg(price),2) as 'Average Price'
from TripData
where lunch_hours == 1

 * sqlite://
Done.


[(16.5,)]

In [ ]:
%%sql
select round(avg(price),2) as 'Average Price'
from TripData
where afternoon == 1

 * sqlite://
Done.


[(16.53,)]

In [ ]:
%%sql
select round(avg(price),2) as 'Average Price'
from TripData
where rush_hours == 1

 * sqlite://
Done.


[(16.56,)]

In [ ]:
%%sql
select round(avg(price),2) as 'Average Price'
from TripData
where night == 1

 * sqlite://
Done.


[(16.56,)]

Average prices spike in 
- Rush hours [4:00 pm - 8:00 pm]
- Night hours [8:00 pm - 5:00 am]
- Early morning hours [5:00 am - 10:00 am]

Lowest average prices are recorded:
- Late morning hours [10:00 am - 12:00 pm]


In [ ]:
# Find out which days have the most trips
%%sql
select day_x as 'Day', count() as 'Number of Trips'
from TripData
group by day_x
order by count() DESC;

 * sqlite://
Done.


[('Thursday', 255983),
 ('Wednesday', 250642),
 ('Tuesday', 214821),
 ('Monday', 184053),
 ('Saturday', 89036),
 ('Sunday', 85715),
 ('Friday', 84746)]

### Results of Analysis

After analysis, cloud coverage has been determined to be a notable factor in whether there is a price surge or not. Trips with a price surge of x 3 have an average cloud cover of 0.8 

The time of day has an impact on the price surges as well, with trips occuring during early mornings and night hours having a higher average pricing as well as a higher percentage of trips with surges; 3.43% and 3.33% respectively. An interesting discovery is that lunch hour trips have the highest percentage of trips with surges; 3.59%

The prices on weekends and weekends are relatively equal, however, the days of the week that have the highest average price are weekdays, Tuesday for Uber and Thursday for Lyft. The days with the highest number of trips made are Thursdays and Wednesdays